metaquot/ppxlib_metaquot.ml

```diff
@@ -128,9 +233,18 @@ module Patt = Make (struct
 
   let annotate p core_type = ppat_constraint ~loc:core_type.ptyp_loc p core_type
 
-  let cast ext =
+  let cast self ext attrs ~type_name =
     match snd ext with
-    | PPat (p, None) -> p
+    | PPat (p, None) ->
+        (match attrs with
+        | None -> ()
+        | Some { quoted_attributes; field_name = _ } ->
+            (* We can't construct a pattern that searches for [quoted_attributes]
+               at the end of [p]'s attribute list -- the pattern language isn't
+               expressive enough. Instead, we fail.
+            *)
```

Comment:
I'm not sure about this comment...
I think the reason for refusing attributes here, is that we are already refusing any attribute in a pattern quotation (see [line 62](https://github.com/ocaml-ppx/ppxlib/pull/441/files#diff-f0e37284b1140251946b048782d09f22023de852392a992294464806eb4fc9afR62) of your version).
For instance `let [%expr 1 [@attr]] = [%expr 1 [@attr]]` will fail, because of the attribute in the pattern extension node.
I guess the reason for this is that taking the attribute into account when making a pattern would make the pattern a bit too restrictive.

However, I don't think there is any technical impossibility with this, it's only a feature we want. We could accept attributes and make the pattern "attribute-sensitive". In this case, we could always build at compile-time the pattern with the merged attributes:
```
let [%expr [%e? { pexp_attributes = [ x ] ; _ }] [@attr2]] = [%expr 1 [@attr1] [@attr2]]
```
would have `x` be the `attr1` attribute.

The reason we can do this (despite the fact that the language of patterns is very limited) is that since patterns are not first class OCaml object (unlike expressions), there cannot be any variable containing a pattern. So, we always have the concrete pattern that we can examine. For instance this has no "pattern" equivalent:

```
let f x =
  [%expr [%e x [@attr1]] [@attr2]]
```

Reply:
My concrete suggestion is just to remove the comment:
```suggestion
```
or to modify it along the lines of:
```suggestion
            (* We don't consider attributes in pattern quotations, since that
               would make the pattern too restrictive. We thus require the
               quotation payload to be free from attributes. *)
```
(feel free to improve!)

Reply:
It may well be true that some patterns are "mergeable", like `[%e? { pexp_attributes = [ x ]; _ }]`. But I think this only works for literal list patterns. For example, I can't think of a way that this could work:
```
let [%expr [%e? { pexp_attributes = x; _ }] [@attr3]] = [%expr 1 [@attr1] [@attr2] [@attr3]]
```

`[%expr [%e? { pexp_attributes = x; _ }] [@attr3]]` would need to expand to a pattern that matches the *prefix* of a list to `x` and its last element to `[@attr3]`. I can't see how this could work: OCaml's pattern language isn't expressive enough for this.

That said, I agree the current comment isn't clarifying so I will delete it.

Reply:
Thanks, that explains it! You are right.
I still think that it is a good thing to remove the comment, since (even if we could) we don't want to merge attributes on patterns.